grafana/grafana
# CloudWatch: Deprecate dynamic labels feature toggle, remove support for Alias in backend

**What is this feature?**

In the backend of the CloudWatch plugin, this 
* Removes any check for the Dynamic Label feature toggle, meaning Dynamic Label can no longer be disabled
   * setting the feature toggle off (cloudWatchDynamicLabels = false) will no longer have any effect
* Removes support for the Alias field to be used directly
* Still migrates Alias to Dynamic Label once: The migration will be handled on the fly, and will be propagated to the dashboard json model only when the user opens the query editor. Once the user has saved the dashboard model with the migrated label value, migration won't happen again. This also applies to Alert rules.

The change of the frontend code will be in a separate PR.
The removal of the feature toggle itself will be in a separate PR.

**Why do we need this feature?**

This was slated for removal in Grafana 10

**Which issue(s) does this PR fix?**: 



Contributes to https://github.com/grafana/grafana/issues/66437

**Special notes for your reviewer:**
Alias has been replaced by Dynamic Label in https://github.com/grafana/grafana/issues/48434

Testing notes:

Normal queries:
* Start on the main branch with the cloudWatchDynamicLabels toggle off
```
[feature_toggles]
cloudWatchDynamicLabels = false
```
* (Still on main branch) Create a dashboard with a CloudWatch datasource panel
* You should see the Alias field. Enter some former Alias pattern like `{{period}}`
* Save that dashboard.
* Now you can check out this PR and restart Grafana. 
* No matter if the feature toggle is false or true, you should see no longer see the Alias field. It should now be a Label field populated with a migrated pattern, e.g. ${PROP('Period')}
   * This Label will not be saved unless the dashboard is saved.  
   * Alert rules with Alias will not have their Label saved unless the rule is edited (opened and viewed) and saved again.
   
In alerts:
* Create a dashboard with a CloudWatch datasource panel.
* In the dashboard JSON model, do not include a `label` target in the panel yet, just use an `alias` target like:
```
      "targets": [
        {
          "alias": "{{period}}",
          "datasource": {
            "type": "cloudwatch",
            "uid": "xpyRJd9Mz"
          },
          "dimensions": {},
          "expression": "",
          "id": "",
          "logGroups": [],
          "matchExact": true,
          "metricEditorMode": 0,
          "metricName": "CPUUtilization",
          "metricQueryType": 0,
          "namespace": "AWS/EC2",
          "period": "",
          "queryMode": "Metrics",
          "refId": "A",
          "region": "default",
          "sqlExpression": "",
          "statistic": "Average"
        }
      ],
```
This simulates a "legacy alias" query.
* Create an Alert based on this query. 
* With this PR, regardless of cloudWatchDynamicLabels equal to true or false, we should see requests from the Alert query being made to the GetMetricData API with `label`. (In this example, alias would have been migrated to label equal to `${PROP('Period')}`)
   * this assertion was made in unit tests and locally by spying on the value of label while running (e.g. added a log line, can also check it by debugging with a breakpoint right around [here](https://github.com/grafana/grafana/blob/d7bd06a87ea73e6dddd3092465a1bdb0d4598dfa/pkg/tsdb/cloudwatch/metric_data_query_builder.go#L24))

Please check that:
